### PR TITLE
Introduce `tanzu local` model and deprecate `tanzu standalone-cluster`

### LINKED ISSUE #2266: Offer minimal deployment model that supports development and experimentation of Tanzu
## Asks

1. Read this proposal
2. [Try out](https://github.com/vmware-tanzu/community-edition/blob/standalone-overhaul/cli/cmd/plugin/standalone-cluster/README.md) the proposed model
3. Vote 👍  or 👎  on this issue
4. If you have additional feedback, respond to this issue

We aim to `accept` or `reject` this proposal 60 days after opening (`12/17/2021`).

## Proposal

⚠️ The name of this feature is currently being determined, you may see it referred to as `local`, `standalone`, and `unmanaged-cluster` in various places. We have decided to move forward with the name `unmanaged-cluster`. Please note that references to `local` or `standalone` (not `standalone-cluster`) represent `unmanaged-cluster`

🚨  This proposal has been partially implemented to help further the conversation around whether we should accept it in this project. [Read here](https://github.com/vmware-tanzu/community-edition/blob/standalone-overhaul/cli/cmd/plugin/standalone-cluster/README.md) for details on how to try it and design details. 🚨

Standalone clusters (SAC) are our attempt to provide workload clusters without the need of a long-running management cluster. With this, we intended to:

* Minimize resource requirements for clusters
* Reduce on boarding time from download to cluster

To accomplish this, we re-purposed Cluster API and extended TKG-Lib (via tanzu-framework) to create the standalone cluster model. With this model in use for many months we've learned that:

1. SAC users often wanted a single-node or local cluster.
1. Users wanting more complex cluster-lifecycle management, need managed clusters.
1. Our usage of cluster-api and tkg-lib created a poor ux for `1` (above) and a stunted experience for `2` (above).
1. Maintaining the functionality support in our dependencies has had significant overhead as it's a use case tangential to our dependencies.

We believe that solving `1` is high-value for those using Tanzu.  We also believe that attempting to replicate cluster-lifecycle management on a single node comes at an inappropriate cost (via dependencies).

**We propose the deprecation of `standalone-cluster` in favor of introducing `local` (clusters).**

## High-level implementation details

In Tanzu, a Management Cluster does the processing of a `TanzuKubernetesRelease` (TKR). It uses the TKR to determine how to create a workload cluster.  

In the `local` model, we'll move the management cluster’s TKR processing client-side. After processing the TKR, we have all the information needed to create a Tanzu [workload] cluster that looks similar to that of one created by a management-cluster. See the following depiction of this relationship.

image

As seen above, after parsing the TKR (client-side) and understanding properties of the to-be-created cluster, we can call into a local provider to create a minimal cluster. By leveraging a provider abstraction (interface), we can insulate ourselves from the underlying details of how the infra/host/desktop-env are created. What matters is we receive a kubeconfig with admin access to the API server.

Our initial provider implementation will be `kind` because it's widely accepted in the Kubernetes community. The following gif demonstrates the bootstrap UX for a local Tanzu cluster.

_GIFs are broken up to save file-size_

### Cluster list / deletion:

`local` is good for:

* Workstation use cases of Tanzu
* Minimal (non-prod) resourced use cases of Tanzu
* CI/CD validations atop Tanzu
* Validation of newly created TKRs

`local` is not meant for:

* Cluster-lifecycle simulation
  * For this, use `management-cluster`.

Additionally, this approach would inherently solve many issues we face today:

* Resource utilization would be **drastically reduced**.
* WSL/Docker Desktop could be natively supported.
* Workstation/Host restarts would be supported.
* Added support for running multiple clusters.
* Testing of different component versions (e.g. kapp-controller) would be possible.
* We would delete our fork of `cluster-api` and `tanzu-framework`.

For in-depth implementation details, [please see our PR](https://github.com/vmware-tanzu/community-edition/pull/2257).

## Release Plan

* `0.10.0` will feature this new model alongside the existing `standalone-cluster` model.
* In `0.10.0`, the existing `standalone-cluster` model will print a deprecation notice to the user.
* In `0.11.0`, we'll remove the existing `standalone-cluster` model.

## FAQ

_This section will be updated as questions come in_

* **Q: What about non-local standalone-clusters (AWS, Azure, vSphere)**
  * A: Execution of this proposal **will** cause a gap for AWS, Azure, and vSphere as there will no longer be non-managed clusters available. However, `standalone-cluster`s are essentially very-limited management clusters with a few components ripped out. For users wanting to test and deploy a *single* cluster in one of those environments, we encourage simply creating a management-cluster and scheduling workloads to it. This **is not** our production-ready advise, but can get you the exact functionality (plus some) of the existing standalone-cluster model.

### PR DESCRIPTION
**Pull request by joshrosso**
**Warning**:  The issue this PR correlates to (https://github.com/vmware-tanzu/community-edition/issues/2266) is under review. We have not committed to this functionality at this point. For discussion around the proposal **please** do this inside the issue and keep PR comments to the implementation details. Thank you!

This PR implements `tanzu local` as an alternative to `standalone-clusters`.

### Testing it:

Right now, several assumptions are made about your environment. Namely that you have tanzu (0.9.0+) pre-installed. If so, you can run the following to try out `tanzu local`

1. Be sure to run all command **from the root** of this repository

    ```sh
     make build-tce-cli-plugins ENVS=linux-amd64 && cp -v ./artifacts/local/v0.10.0-dev.1/tanzu-local-linux_amd64 ~/.local/share/tanzu-cli/tanzu-plugin-local
    ```
    
    > If not on `linux`, change the above to point to your `$XDG_DATA_HOME`

1. Run `tanzu local`
    
    ```
    $ tanzu local
    Manage local environments of Tanzu
    
    Usage:
      tanzu local [command]
    
    Available Commands:
      completion    generate the autocompletion script for the specified shell
      create        create a local tanzu cluster
      delete        delete a local tanzu cluster
      list          list tanzu clusters
    
    Flags:
      -h, --help              help for local
          --log-file string   Log file path
      -v, --verbose int32     Number for the log level verbosity(0-9)
    
    Use "tanzu local [command] --help" for more information about a command.
    ```

### Cluster list / deletion:

Resolves: https://github.com/vmware-tanzu/community-edition/issues/2266

### CODE REVIEW

**Comment by jorgemoralespou** [comment]
>> make build-tce-cli-plugins ENVS=linux-amd64 && cp -v ./artifacts/local/v0.10.0-dev.1/tanzu-local-linux_amd64 ~/.local/share/tanzu-cli/tanzu-plugin-local

>If not on linux, change the above to point to your $XDG_DATA_HOME

File: cli/cmd/plugin/standalone-cluster/create.go [hunk 1]
```diff
@@ -5,17 +5,42 @@ package main
 
 import (
 	"fmt"
+	"io/ioutil"
```
Comment (nrb) [inline cli/cmd/plugin/standalone-cluster/create.go:8]: I think this directory should be renamed, for clarity.

File: cli/cmd/plugin/standalone-cluster/create.go [hunk 2]
```diff
@@ -54,103 +83,374 @@ func create(cmd *cobra.Command, args []string) error {
...
+				if err != nil {
+					return err
+				}
+
+				docBytes, err := docSet.AsBytes()
+				if err != nil {
+					return err
+				}
+
+				// TODO(jpmcb): in the future, maybe we consider using cliient-go to actually create the objects instead of shilling out to kubectl
```
Comment (nrb) [inline cli/cmd/plugin/standalone-cluster/create.go:304]: +1 to client-go.

**Review by nrb — COMMENTED** [review]
Looks like `cli/cmd/plugin/standalone-cluster/standalone-cluster` is a committed binary, not sure if you wanted to do that.

Also, is there a definition of the provider interface we're aiming to create as mentioned in the issue yet? I'm not seeing one by searching for `interface` if so.

**Pull request closed without merging**